opm/ml/keras_model.cpp

```diff
@@ -0,0 +1,419 @@
+    KASSERT(ReadUnsignedInt(file, &weights_cols), "Expected weight cols");
+    KASSERT(weights_cols > 0, "Invalid weights shape");
+
+    weights_.Resize(weights_rows, weights_cols);
+    KASSERT(
+        ReadFloats(file, weights_.data_.data(), weights_rows * weights_cols),
+        "Expected weights");
+
+    return true;
+}
+
+template<class Evaluation>
+bool KerasLayerEmbedding<Evaluation>::Apply(Tensor<Evaluation>* in, Tensor<Evaluation>* out) {
+    int output_rows = in->dims_[1];
+    int output_cols = weights_.dims_[1];
+    out->dims_ = {output_rows, output_cols};
+    out->data_.reserve(output_rows * output_cols);
+
+    std::for_each(in->data_.begin(), in->data_.end(), [=](Evaluation i) {
+        typename std::vector<Evaluation>::const_iterator first =
+            this->weights_.data_.begin() + (getValue(i) * output_cols);
+        typename std::vector<Evaluation>::const_iterator last =
+            this->weights_.data_.begin() + (getValue(i) + 1) * output_cols;
+
+        out->data_.insert(out->data_.end(), first, last);
+    });
+
+    return true;
+}
+
+
+template<class Evaluation>
+bool KerasModel<Evaluation>::LoadModel(const std::string& filename) {
+    std::ifstream file(filename.c_str(), std::ios::binary);
+    KASSERT(file.is_open(), "Unable to open file %s", filename.c_str());
+
+    unsigned int num_layers = 0;
+    KASSERT(ReadUnsignedInt(&file, &num_layers), "Expected number of layers");
+
+    for (unsigned int i = 0; i < num_layers; i++) {
+        unsigned int layer_type = 0;
+        KASSERT(ReadUnsignedInt(&file, &layer_type), "Expected layer type");
+
+        KerasLayer<Evaluation>* layer = NULL;
+
+        switch (layer_type) {
+        case kFlatten:
+            layer = new KerasLayerFlatten<Evaluation>();
+            break;
+        case kScaling:
+            layer = new KerasLayerScaling<Evaluation>();
+            break;
+        case kUnScaling:
+            layer = new KerasLayerUnScaling<Evaluation>();
+            break;
+        case kDense:
+            layer = new KerasLayerDense<Evaluation>();
+            break;
+        case kActivation:
+            layer = new KerasLayerActivation<Evaluation>();
+            break;
+        default:
+            break;
+        }
+
+        KASSERT(layer, "Unknown layer type %d", layer_type);
+
+        bool result = layer->LoadLayer(&file);
+        if (!result) {
+            printf("Failed to load layer %d", i);
+            delete layer;
+            return false;
+        }
+
+        layers_.push_back(layer);
+    }
+
+    return true;
+}
+
+template<class Evaluation>
+bool KerasModel<Evaluation>::Apply(Tensor<Evaluation>* in, Tensor<Evaluation>* out) {
+    Tensor<Evaluation> temp_in, temp_out;
+
+    for (unsigned int i = 0; i < layers_.size(); i++) {
+        if (i == 0) {
+            temp_in = *in;
+        }
+
+        KASSERT(layers_[i]->Apply(&temp_in, &temp_out),
+                "Failed to apply layer %d", i);
+
+        temp_in = temp_out;
```

Comment:
Implement an O(1) swap() function instead, and avoid the copying in the layer apply() functions.

Reply:
Why? Do you see a compile error, or wrong runtime behaviour?

Reply:
Wrong runtime behaviour for std::move and wrong values for std::swap.

Reply:
Please show your swap() implementation, maybe we can figure out what is wrong.

Reply:
template<class Evaluation>
bool NNModel<Evaluation>::apply(Tensor<Evaluation>& in, Tensor<Evaluation>& out) {
    Tensor<Evaluation> temp_in, temp_out;

    for (unsigned int i = 0; i < layers_.size(); i++) {

            OPM_ERROR_IF(!(layers_[i]->apply(in, out)),fmt::format("\n Failed to apply layer " "{}", i));
            
            std::swap(out,in);
}
    return true;

}